[sendtohttp]  Add Open-Meteo events + JSON event

chromoxdor opened the pull request:
To-do:
- [x] Lots of documentation and examples
- [x] Some more annotation in the code probably 

___
As being asked here: https://www.letscontrolit.com/forum/viewtopic.php?p=72083#p72083 i do a PR for Open-Meteo events

1. It parses url for the variable type (current,hourly, daily)
2.  Contrains the response and parses this for the variables
3. depending on the variable type it generates an event (e.g. `On OpenMeteo#current Do...` )

It works but responses can get quite long, especially for hourly variables and also when all types combined together.
But for now it was working fine on an esp32 classic even with very long resonses > 6kb 

I tried to kind of create a size check but i am no expert for memory constraint systems. Maybe you can help out here with some recommendations.

I guess when calling this it is already to late 🙂 :
```cpp
 if (uri.length() > 5000) {
        addLog(LOG_LEVEL_ERROR, F("Response exceeds 5000 characters"));
      }
      else {
```

**How to use:** 
1. visit https://open-meteo.com/en/docs
2.  click together what you need and scroll down to the API Response section where you find this:
<img width="526" alt="Bildschirmfoto 2025-01-02 um 14 59 15" src="https://github.com/user-attachments/assets/75ea94fb-89ed-4659-88c8-93de245613da" />

3. use`sendtohttp api.open-meteo.com,80,/v1/forecast?...`

**Issues:**
1.  ~~ESPEasy throws a "Too many arguments" error  but it still works fine~~
2.  In the log only 25 eventvalues are displayed. But you can still access all of them  with rules. I wonder where the limit is?

chromoxdor committed:
Update Networking.cpp

chromoxdor commented:
Command:
`sendtohttp api.open-meteo.com,80,"/v1/forecast?latitude=52.5244&longitude=13.4105&current=temperature_2m,relative_humidity_2m&daily=temperature_2m_max,temperature_2m_min,uv_index_max&timezone=Europe%2FLondon&forecast_days=3"
`

Response:
```
2580494: EVENT: OpenMeteo#current=4.3,76
2580503: EVENT: OpenMeteo#daily=7.5,2.4,1.2,1.5,-1.0,-1.1,1.00,0.95,0.70
```

tonhuisman commented:
> 1. ESPEasy throws a "Too many arguments" error  but it still works fine

Wrap the (long) url, that contains several commas, in quotes, and this issue should be resolved 😉 
Alternative: url-encode the commas to `%2C` so ESPEasy doesn't see them as commas

chromoxdor commented:
> Wrap the (long) url, that contains several commas, in quotes, and this issue should be resolved 😉

I tried that and it didn't work...but now i think i know what happend. I have this auto correction plugin in my browser. Disabling it helped :)

chromoxdor committed:
a tiny bit better formatting

tonhuisman commented on src/src/Helpers/Networking.cpp:
```diff
@@ -1685,11 +1685,11 @@ int http_authenticate(const String& logIdentifier,
 
     # if FEATURE_OMETEO_EVENT
```
IMHO, a somewhat better name for this would be `FEATURE_OPENMETEO_EVENT`, ometeo is a bit ambiguous, I think.

Also, this variable, and a small summary of the description, should go in the `Custom-sample.h` file, and a check with default into `define_plugin_sets.h`:
```
#ifndef FEATURE_OPENMETEO_EVENT
  #define FEATURE_OPENMETEO_EVENT 0 // Or maybe 1 for ESP32 (MAX) builds
#endif
```

chromoxdor replied:
I agree..

tonhuisman commented on src/src/Helpers/Networking.cpp:
```diff
@@ -1682,6 +1682,126 @@ int http_authenticate(const String& logIdentifier,
+        String currentParams = getCombinedParams(uri, "current");
+        String hourlyParams  = getCombinedParams(uri, "hourly");
+        String dailyParams   = getCombinedParams(uri, "daily");
+
+        auto processAndQueueParams = [](String& params, String& str, String eventName) {
+                                       if (!params.isEmpty()) {
+                                         String keys[20];
+                                         int    keyCount = 0;
```
Is this part formatted using Uncrustify? (because it doesn't look that way...)

chromoxdor replied:
I have some issues with uncrustify... in the beginning I formatted the whole document. But it changed a lot so i reverted it. Then i tried to only format the selection. But then the selected code disappears completely 🤷‍♂️
I was to lazy to take care of that so formatting is a bit off....

tonhuisman replied:
AFAICS, formatting only a section from VSCode doesn't work, so I always format the entire document 😺 
When changing a small part of the code, and Uncrustify would reformat a lot of the file, I format the document, copy the formatted part I want to keep, undo the formatting and paste back the part I want to keep. As a work-around for the not working 'format selection'.

chromoxdor replied:
Ok... so i am not the only one with the issue. Thanks for the tip :)

chromoxdor replied:
I did as you asked but it did not change so much....

chromoxdor committed:
changes for OPENMETEO_EVENT and THINGSPEAK_EVENT

- added FEATURE_OPENMETEO_EVENT to Custom-sample.h
- disabled FEATURE_THINGSPEAK_EVENT and FEATURE_OPENMETEO_EVENT on default
- both got the same structure in define_plugin_sets.h

TD-er commented on src/src/Helpers/Networking.cpp:
```diff
@@ -1682,6 +1682,125 @@ int http_authenticate(const String& logIdentifier,
+        addLog(LOG_LEVEL_ERROR, F("Response exceeds 5000 characters"));
+      }
+      else {
+        String str = http.getString();
```
Please use `const` where possible.
At first glance, `str` is not changed anymore, so it should be `const String str`.

TD-er commented on src/src/Helpers/Networking.cpp:
```diff
@@ -1682,6 +1682,125 @@ int http_authenticate(const String& logIdentifier,
+        String str = http.getString();
+
+        // Process URL to get unique keys
+        auto getCombinedParams = [](String url, String paramName) {
```
Use `const String& url, const String& paramName` as this prevents copying the arguments on each call.

TD-er commented on src/src/Helpers/Networking.cpp:
```diff
@@ -1682,6 +1682,125 @@ int http_authenticate(const String& logIdentifier,
+
+        // Process URL to get unique keys
+        auto getCombinedParams = [](String url, String paramName) {
+                                   int start = url.indexOf(paramName + "=");
```
Better use `concat(paramName, '=')` as argument, will (hopefully) result in smaller build size.

TD-er commented on src/src/Helpers/Networking.cpp:
```diff
@@ -1682,6 +1682,125 @@ int http_authenticate(const String& logIdentifier,
+        auto getCombinedParams = [](String url, String paramName) {
+                                   int start = url.indexOf(paramName + "=");
+
+                                   if (start == -1) { return String(""); }
```
We have a global defined `EMPTY_STRING` for this purpose, so you can call `return EMPTY_STRING;`

TD-er commented on src/src/Helpers/Networking.cpp:
```diff
@@ -1682,6 +1682,125 @@ int http_authenticate(const String& logIdentifier,
+
+                                   if (start == -1) { return String(""); }
+                                   start += paramName.length() + 1;
+                                   int end = url.indexOf('&', start);
```
`end` does not seem to be changed later, so better make it `const int end...`

TD-er commented on src/src/Helpers/Networking.cpp:
```diff
@@ -1682,6 +1682,125 @@ int http_authenticate(const String& logIdentifier,
+        // Extract current hourly and daily parameters
+        String currentParams = getCombinedParams(uri, "current");
+        String hourlyParams  = getCombinedParams(uri, "hourly");
+        String dailyParams   = getCombinedParams(uri, "daily");
```
Please use the `F()` macro and the String objects should be `const`
```c++
const String dailyParams   = getCombinedParams(uri, F("daily"));
```

TD-er commented on src/src/Helpers/Networking.cpp:
```diff
@@ -1682,6 +1682,125 @@ int http_authenticate(const String& logIdentifier,
+        String hourlyParams  = getCombinedParams(uri, "hourly");
+        String dailyParams   = getCombinedParams(uri, "daily");
+
+        auto processAndQueueParams = [](String& params, String& str, String eventName) {
```
At first glance, it looks like the arguments are not changed, so you better use:
```c++
const String& params, const String& str, const String& eventName
```

TD-er commented on src/src/Helpers/Networking.cpp:
```diff
@@ -1682,6 +1682,125 @@ int http_authenticate(const String& logIdentifier,
+                                         // Split and add keys to the array
+                                         while (commaIndex != -1) {
+                                           String key = params.substring(startIndex, commaIndex);
+                                           keys[keyCount++] = key;
```
I don't see any check for max. `keyCount` and since it is a hard limited array, you should make sure it never gets out of bounds.

TD-er commented on src/src/Helpers/Networking.cpp:
```diff
@@ -1682,6 +1682,125 @@ int http_authenticate(const String& logIdentifier,
+
+                                         // Add the last key
+                                         String lastKey = params.substring(startIndex);
+                                         keys[keyCount++] = lastKey;
```
Ditto, missing bounds check.

TD-er commented on src/src/Helpers/Networking.cpp:
```diff
@@ -1682,6 +1682,125 @@ int http_authenticate(const String& logIdentifier,
+                                         }
+
+                                         // Add the last key
+                                         String lastKey = params.substring(startIndex);
```
Looks like it can be a `const String lastKey`

TD-er commented on src/src/Helpers/Networking.cpp:
```diff
@@ -1682,6 +1682,125 @@ int http_authenticate(const String& logIdentifier,
+                                         String lastKey = params.substring(startIndex);
+                                         keys[keyCount++] = lastKey;
+
+                                         String csv              = "";
```
No need to initialize with an empty `""` as the default `String` constructor already is an empty string.

TD-er commented on src/src/Helpers/Networking.cpp:
```diff
@@ -1682,6 +1682,125 @@ int http_authenticate(const String& logIdentifier,
+
+                                         String csv              = "";
+                                         int    startStringIndex = str.indexOf("\"" + eventName + "\":") + eventName.length() + 4;
+                                         int    endStringIndex   = str.indexOf("}", startStringIndex);
```
These two look like they can also be `const int`

TD-er commented on src/src/Helpers/Networking.cpp:
```diff
@@ -1682,6 +1682,125 @@ int http_authenticate(const String& logIdentifier,
+                                           if (startIndex == -1)
+                                           {
+                                             // Handle case where key is not found
+                                             value = "-256"; // Placeholder value
```
Please use `F()` macro

TD-er commented on src/src/Helpers/Networking.cpp:
```diff
@@ -1682,6 +1682,125 @@ int http_authenticate(const String& logIdentifier,
+                                           {
+                                             int endIndex = 0;
+
+                                             if (eventName != "current") { // in daily and hourly the values are stored in an
```
Please use `F()` macro

tonhuisman replied:
Or even better:
`if (!equals(eventName, F("current"))) {`

TD-er commented on src/src/Helpers/Networking.cpp:
```diff
@@ -1682,6 +1682,125 @@ int http_authenticate(const String& logIdentifier,
+                                           }
+                                           csv += value;
+                                         }
+                                         eventName = "OpenMeteo#" + eventName;
```
Please use `concat(F("OpenMeteo#"), eventName)`

TD-er commented on src/src/Helpers/Networking.cpp:
```diff
@@ -1682,6 +1682,125 @@ int http_authenticate(const String& logIdentifier,
+
+        processAndQueueParams(currentParams, str, "current");
+        processAndQueueParams(hourlyParams,  str, "hourly");
+        processAndQueueParams(dailyParams,   str, "daily");
```
Please use F() macro on these C-style strings

TD-er commented on src/src/Helpers/Networking.cpp:
```diff
@@ -1682,6 +1682,125 @@ int http_authenticate(const String& logIdentifier,
+
+                                           if (!csv.isEmpty())
+                                           {
+                                             csv += ",";
```
```c++
csv += ',';
```
Will result in smaller code as you only need to deal with a char instead of calling a complete constructor of a String using a C-style string.

chromoxdor committed:
Update Networking.cpp

chromoxdor commented:
Thanks for helping me...

chromoxdor committed:
combining two functions

chromoxdor commented:
I combined the two functions because there is actually no need for that

chromoxdor committed:
wrong use of concat times two :)

TD-er commented:
> I combined the two functions because there is actually no need for that

Yep good fix :)

tonhuisman commented on src/src/CustomBuild/define_plugin_sets.h:
```diff
@@ -3621,11 +3621,11 @@ To create/register a plugin, you have to :
-    #endif
+  #endif
+
+  #ifndef FEATURE_OPENMETEO_EVENT
+    #define FEATURE_OPENMETEO_EVENT 0
   #endif
```
Would be nice to have both feature flags enabled in the MAX builds 😃

chromoxdor replied:
True! I just thought of it when I realized I was posting nonsense on the forum 😁.
Even though not many people use the max builds I guess...

tonhuisman replied:
> not many people use the max builds I guess...

We _are_ trying to push more users to MAX builds... 😎

tonhuisman replied:
Oh, and maybe they are also useful in the ESP32 Climate builds?

tonhuisman commented on src/src/Helpers/Networking.cpp:
```diff
@@ -1682,6 +1682,115 @@ int http_authenticate(const String& logIdentifier,
+                }
+
+                String csv;
+                const int startStringIndex = str.indexOf("\"" + eventName + "\":") + eventName.length() + 4;
+                const int endStringIndex = str.indexOf("}", startStringIndex);
```
That could better be like:
```
const int startStringIndex = str.indexOf(strformat(F("\"%s\":"), eventName.c_str())) + eventName.length() + 4;
const int endStringIndex   = str.indexOf('}', startStringIndex);
```

tonhuisman commented on src/src/Helpers/Networking.cpp:
```diff
@@ -1682,6 +1682,115 @@ int http_authenticate(const String& logIdentifier,
+                        if (!equals(eventName, F("current"))) {
+                            // In daily and hourly, the values are stored in an array
+                            startIndex += key.length() + 3; // Move index past the key
+                            endIndex = str.indexOf("]", startIndex);
```
Char instead of C string...

tonhuisman commented on src/src/Helpers/Networking.cpp:
```diff
@@ -1682,6 +1682,115 @@ int http_authenticate(const String& logIdentifier,
+                            endIndex = str.indexOf("]", startIndex);
+                        } else {
+                            startIndex += key.length() + 2; // Move index past the key
+                            endIndex = str.indexOf(",", startIndex);
```
Ditto

tonhuisman commented on src/src/Helpers/Networking.cpp:
```diff
@@ -1682,6 +1682,115 @@ int http_authenticate(const String& logIdentifier,
+                        }
+
+                        // Find the index of the next comma
+                        if ((endIndex == -1) || (endIndex > str.indexOf("}", startIndex))) {
+                            endIndex = str.indexOf("}", startIndex); // If no comma is found or comma comes after },
```
Ditto

chromoxdor replied:
This too i guess right?:
`int startIndex = str.indexOf(key + "\":", startStringIndex);`

tonhuisman replied:
Here you can use `...strformat(F("%s\":"), key.c_str())...`

tonhuisman replied:
or `...concat(key, F("\":"))...`

tonhuisman commented on src/src/Helpers/Networking.cpp:
```diff
@@ -1682,6 +1682,115 @@ int http_authenticate(const String& logIdentifier,
+            const String params = (end == -1) ? url.substring(start) : url.substring(start, end);
+
+            if (!params.isEmpty()) {
+                String keys[20];
```
Please try to avoid using so called magic numbers.
Better to use a `#define WEATHER_KEYS_MAX 20` or something similar, so it can be managed from 1 place in the source.
An alternative could be to use a `std::vector` instead of an array, so it will grow as needed 😉 

NB: Common exceptions for magic numbers are -1, 0 and 1.

chromoxdor replied:
what about: 
```
if (uri.length() > 5000) {
        addLog(LOG_LEVEL_ERROR, F("Response exceeds 5000 characters"));
      }
      else {
```
This is not really necessary right?

tonhuisman replied:
Well, the max. size could very well be dependent on the MCU generation, for ESP8266 is might even fail on smaller requests...

chromoxdor replied:
but as soon as i call `uri.length() ` it will be in ram and therefore i prevented nothing with the statement? At least that's what I imagine

tonhuisman replied:
If it is too large to fit in memory, the ESP will crash before reaching your check 💥

chromoxdor replied:
so i could leave it as a warning message process the rest anyway when esp is still running at this point :)

chromoxdor committed:
update

- add events to max builds
- minor code changes
- size warnings: defines added

tonhuisman commented on src/src/Helpers/Networking.cpp:
```diff
@@ -1695,101 +1695,104 @@ int http_authenticate(const String& logIdentifier,
+
+      const String str = http.getString();
+
+      if (str.length() > 5000) {
```
You missed a spot 😸

chromoxdor committed:
Update Networking.cpp

chromoxdor commented:
By the way @tonhuisman I have the feeling that we already have a JSON parser in ESPEasy somewhere...
Can this be?

chromoxdor commented:
@tonhuisman , @TD-er 
Another question:
Would it make sense to move these "special" parser into another file?

TD-er commented:
> @tonhuisman , @TD-er Another question: Would it make sense to move these "special" parser into another file?

It would at least make the code more readable.
You can place the .h and .cpp files in the Helpers folder.
Make sure to only include those files only if the feature is enabled, so you don't link in unneeded code when the feature is not used.

tonhuisman commented:
> By the way @tonhuisman I have the feeling that we already have a JSON parser in ESPEasy somewhere... Can this be?

Yes, we have, but it's not easy to use when parsing unknown data (see P037 code... 😞).

chromoxdor committed:
put thingspeak and openmeteo events into extra file

tonhuisman commented on src/src/CustomBuild/define_plugin_sets.h:
```diff
@@ -3636,6 +3636,12 @@ To create/register a plugin, you have to :
     #endif
   #endif
 
+  #if defined(FEATURE_THINGSPEAK_EVENT) || defined(FEATURE_OPENMETEO_EVENT)
```
These should check for the defines to be true/1, when 0 they are also defined...
`#if FEATURE_THINGSPEAK_EVENT || FEATURE_OPENMETEO_EVENT`

chromoxdor replied:
Ahh true...

TD-er replied:
> Ahh true...

... or false ;)

chromoxdor commented:
> Yes, we have, but it's not easy to use when parsing unknown data (see P037 code... 😞).

I was thinking of using the Arduino code for parsing, but for now it is not worth it size-wise.

tonhuisman commented:
> I was thinking of using the Arduino code for parsing, but for now it is not worth it size-wise.

We already use ArduinoJSON library (though not the latest 7.x as it got quite some increase in size), that works very well when having known input.
Maybe it works in this case, as the structure is predefined, only the keys are variable, and AFAICS, you can use `element.sub.subsub` (dot) notation to retrieve values from the loaded JSON. You probably need to use `.as<float>()` or something similar (don't have the examples at hand).

TD-er commented on src/src/Helpers/HTTPResponseParser.cpp:
```diff
@@ -0,0 +1,171 @@
+
+                                     String csv;
+                                     const int startStringIndex = str.indexOf(strformat(F("\"%s\":"),
+                                                                                        eventName.c_str())) + eventName.length() + 4;
```
With string operations like these, where you add a fixed offset, it is often a lot simpler to read when you add some example line as comment.

chromoxdor committed:
added inverter event

chromoxdor commented:
I added also the inverter event. It seems that  a bunch of inverter using the same API.
I also stole the JSON code from P037. 😬
Maybe in the near future i will use this code to iterate through the open-meteo keys as well...

chromoxdor committed:
fix for not building on minimal_core_274_esp8266...

chromoxdor commented:
I don't understand why the action runs are failing. For me these builds are working just fine fine...

chromoxdor committed:
#ifdef to #if. m(

TD-er commented on src/src/Helpers/HTTPResponseParser.cpp:
```diff
@@ -0,0 +1,243 @@
+#include "../Helpers/Networking.h"
```
I expected an include of `../Helpers/HTTPResponseParser.h` on the first line.

And then following line to have this:
```c++
#if RESPONSE_PARSER_SUPPORT
```

TD-er commented on src/src/Helpers/HTTPResponseParser.h:
```diff
@@ -0,0 +1,18 @@
+#include "../../_Plugin_Helper.h"
+
+#ifdef PLUGIN_BUILD_MINIMAL_OTA
+# include <ESP8266HTTPClient.h>
```
This include should also be wrapped in `#ifdef ESP8266` .... `#endif`

chromoxdor committed:
more universal json event approach

TD-er commented on src/src/Helpers/HTTPResponseParser.h:
```diff
@@ -3,16 +3,28 @@
 
 #include "../../ESPEasy_common.h"
 #include "../../_Plugin_Helper.h"
+#include "../Helpers/Networking.h"
```
Here (after the include of _Plugin_Helper.h) you should also have a check for `#if RESPONSE_PARSER_SUPPORT`
So there isn't a function declaration being made if there also isn't a function body in the .cpp

chromoxdor commented:
OK, hopefully this time the action run finishes without errors.
I changed the "inverter" function for a more universal approach. 
Therefore i added a function to parse the content of a file called `json.keys` line by line.

example content:
```
Body.Data.DAY_ENERGY.Values.1
Body.Data.PAC.Values.1
```
For the URL: There needs to be &json added to the end of it to tell the parser that there is a JSON-response.
`SendToHTTP 192.168.1.199,80,"/solar_api/v1/GetInverterRealtimeData.cgi?Scope=System?json"`

Actually i would like to see this as an plugin, where you enter your URL and store your keys in the same manner. But that not only mean to create a new plugin but also to deduplicate code because of P037. 
At the moment this is too complicated as the structure is far too complex for a beginner like me.

chromoxdor committed:
Update HTTPResponseParser.h

TD-er commented on src/src/Helpers/HTTPResponseParser.cpp:
```diff
@@ -1,243 +1,337 @@
+      if (value.is<int>()) {
+        csvOutput += String(value.as<int>());
+      } else if (value.is<float>()) {
+        csvOutput += String(value.as<float>());
```
This will always use 2 decimals. Is that good enough?

chromoxdor commented:
> For the URL: There needs to be &json added to the end of it to tell the parser that there is a JSON-response.
> `SendToHTTP 192.168.1.199,80,"/solar_api/v1/GetInverterRealtimeData.cgi?Scope=System?json"`

Usually adding a parameter that is not existent on the server side causes no issues. 
I had no choice for that one since i can not pass a parameter

TD-er commented on src/src/Helpers/HTTPResponseParser.cpp:
```diff
@@ -1,243 +1,337 @@
+        csvOutput += "unknown";
       }
+    } else {
+      csvOutput += "null"; // Indicate missing value
```
This is sent as an argument for an event.
However we cannot match a string type.
Wouldn't it make more sense to just not add anything?
So your string then becomes something like `12,34,,78`
instead of `12,34,null,78`

chromoxdor replied:
Great, i accidentally deleted my replies. I hope you had the chance to read it. Somebody is stressing me to go outside 🤷‍♂️ I rewrite it when i am back....

chromoxdor committed:
added variable decimals for floating point numbers

+ fix typo

chromoxdor committed:
some changes to the json-event

- add array output to the json parser
- add grouping of keys for individual requests

tonhuisman commented on src/src/Helpers/HTTPResponseParser.cpp:
```diff
@@ -235,9 +250,9 @@ void eventFromResponse(const String& host, const int& httpCode, const String& ur
+          readAndProcessJsonKeys(root, numJson);
         } else {
-          Serial.println("Failed to parse JSON");
+          addLog(LOG_LEVEL_INFO, strformat(F("Parsing JSON failed")));
```
No need to use `strformat()` when not having % variables in the message.

tonhuisman commented on src/src/Helpers/HTTPResponseParser.cpp:
```diff
@@ -333,15 +386,18 @@ void readAndProcessJsonKeys(DynamicJsonDocument*root) {
   // Log the results
   addLog(LOG_LEVEL_INFO, strformat(F("Successfully processed %d out of %d keys"), successfullyProcessedCount, keyCount));
-  eventQueue.addMove(strformat(F("JsonReply=%s"), csvOutput.c_str()));
+  eventQueue.addMove(strformat(F("JsonReply%s=%s"),  (numJson != 0) ? String(numJson).c_str() : "", csvOutput.c_str()));
```
Hm, that seems a bit complex.

chromoxdor replied:
it got even a bit more complex :) .. I am just writing a small how-to in the forum ..

chromoxdor replied:
I would be grateful for information on how to make it less complex while retaining its functionality. :)

tonhuisman replied:
We have this function globally available: `String toStringNoZero(int64_t value);` that returns an empty string for a 0 value.

chromoxdor replied:
Could you give me an example with the updated code?

tonhuisman replied:
Can't make it much simpler than this, as a char (`%c`) can't be empty like a string, and you wouldn't want a space there 😕 Changing to Flash strings will explode the code even more 😨.
```
eventQueue.addMove(strformat(F("JsonReply%s%s=%s"),
                               numJson != 0 ? "#" : "",
                               toStringNoZero(numJson).c_str(),
                               csvOutput.c_str()));
```

TD-er replied:
Just to give you an idea about the 'costs' in bin size...

- a char (`'x'` for example) takes 1 byte in the binary as well as in RAM when executing
- A  C-style string `"foo"` takes the number of characters + a 0-byte as termination, both in compiled binary as well as in memory.
- A flash-string (`F("foo")`) takes number of characters + a 0-byte as termination, but will be deduplicated in binary and not copied to RAM. It does however also add a pointer to the string.

If you use a flash-string which needs to be converted to a `String` object, each call will add about 16 bytes as the compiler will wrap a `String` constructor around it.
So there are cases when a C-style string will be the better choice, as those can be used directly in `strformat` calls without the need to wrap it in a `String().c_str()` form.

chromoxdor replied:
Many thanks! That is very helpful and I am trying to internalize this...

chromoxdor committed:
made the eventQueue even more complex :P

chromoxdor committed:
minor changes

-changed variable name for the decimals to decimalsJP

chromoxdor committed:
moved variable to cpp file

tonhuisman commented:
> 2\. In the log only 25 eventvalues are displayed. But you can still access all of them  with rules. I wonder where the limit is?

Is that the web-based log (Tools/Log), or via the serial console?

TD-er commented:
> > 2. In the log only 25 eventvalues are displayed. But you can still access all of them  with rules. I wonder where the limit is?
> 
> Is that the web-based log (Tools/Log), or via the serial console?

Both are truncated to some fixed max. size (though not the same max.)

chromoxdor committed:
updates

- add: log error for json parsing
- change: decimals for float and double reduced by 1 (6,15)
- inceased json buffer

chromoxdor committed:
Update HTTPResponseParser.cpp

unsing double for double now

chromoxdor committed:
Update define_plugin_sets.h

chromoxdor commented:
Should I add a whole new page for documentation of the these parser?
Adding it to the SendToHTTP section is probably way too much information. I would prefer to add a link from there.

chromoxdor committed:
Update HTTPResponseParser.cpp

chromoxdor committed:
Update HTTPResponseParser.cpp

- There is no difference between is<float>() and is<double>() as a value check. Keeping <float>
-#define ARDUINOJSON_USE_DOUBLE 1 does not seem to be needed (only for serializeJson)

chromoxdor commented:
I am inclined to use this for convenience. On a scale of 1-10, how bad is this idea?

```cpp
#  if ESP8266
        root = new (std::nothrow) DynamicJsonDocument(ESP.getMaxFreeBlockSize() - 512);
  #  else // if ESP8266
        root = new (std::nothrow) DynamicJsonDocument(ESP.getMaxAllocHeap());
  #  endif // if ESP8266
```

found here: https://arduinojson.org/v6/how-to/determine-the-capacity-of-the-jsondocument/#technique-4-use-whatever-ram-is-available

Edit: Our scenario would match this: https://arduinojson.org/v6/how-to/determine-the-capacity-of-the-jsondocument/#what-if-i-dont-know-my-input-ahead-of-time

TD-er commented:
It won't even fit on the 0...10 scale.
This is really a bad idea as it will cause lots and lots of issues when creating Strings or any other heap allocation.

chromoxdor committed:
Merge branch 'letscontrolit:mega' into openmeteo-events

chromoxdor committed:
documentation and slight reduction of code

chromoxdor committed:
Update define_plugin_sets.h

chromoxdor committed:
corrected json examples

chromoxdor committed:
Docs: Formatting and linking

chromoxdor committed:
Merge branch 'mega' into openmeteo-events

chromoxdor committed:
fixed typo

chromoxdor committed:
Merge branch 'openmeteo-events' of https://github.com/chromoxdor/ESPEasy into openmeteo-events

chromoxdor committed:
Merge branch 'mega' into openmeteo-events

chromoxdor committed:
Merge branch 'mega' into openmeteo-events

chromoxdor commented:
This is also ready for merging...

chromoxdor committed:
Merge branch 'mega' into openmeteo-events

chromoxdor committed:
Update define_plugin_sets.h

chromoxdor committed:
Merge branch 'mega' into openmeteo-events

chromoxdor commented:
What needs to be done to get this merged?

TD-er commented on src/src/CustomBuild/define_plugin_sets.h:
```diff
@@ -3614,7 +3627,18 @@ To create/register a plugin, you have to :
+    #define FEATURE_STRING_VARIABLES  0
+  #endif
+#endif // ifndef FEATURE_STRING_VARIABLES
+#if FEATURE_STRING_VARIABLES && defined(ESP8266) // NOT supported on ESP8266 because of limited available memory
```
This way you also disable it for ESP8266 custom builds where it was explicitly enabled.

chromoxdor replied:
This is not one of my changes..  :)

chromoxdor replied:
It only looks like it because I was resolving conflicts like the amateur I am. :)

TD-er commented:
> What needs to be done to get this merged?

I will have a look at it now...

TD-er commented on src/src/Helpers/HTTPResponseParser.cpp:
```diff
@@ -0,0 +1,399 @@
+          result = parseStringKeepCase(uri, 4, '/').substring(0, 1) + "," + result; // since the field number is always the fourth part of
+                                                                                    // the
+                                                                                    // url and is always a single digit, we can use this to
+                                                                                    // extact the fieldnumber
```
extract (missing 'r')

TD-er commented on src/src/Helpers/HTTPResponseParser.cpp:
```diff
@@ -0,0 +1,399 @@
+#  endif // ifdef FEATURE_USE_DOUBLE_AS_ESPEASY_RULES_FLOAT_TYPE
+
+  // Open the `json.keys` file
+  const String fileName = strformat(
```
Why using `strformat` here?
You can just construct a string from a flash string.

chromoxdor replied:
good question... its so long a ago i can`t remember :)

chromoxdor commented on docs/source/Rules/HTTPResponseParser.rst:
```diff
@@ -0,0 +1,239 @@
+        LogEntry,'Values : Temp=%eventvalue1% Hum=%eventvalue2% Baro=%eventvalue3% UV=%eventvalue4%'
+    Endon
+
+    //this is a gerneric approacht to grep all the different kinds of weather variables in one rules block
```
found one too :).  "generic approach"

TD-er commented on src/src/Helpers/HTTPResponseParser.cpp:
```diff
@@ -0,0 +1,399 @@
+  String csvOutput;                   // Collect values as a CSV string
+
+  while (keyFile.available()) {
+    String key = keyFile.readStringUntil('\n');
```
What happens if the file doesn't end with a newline?

chromoxdor replied:
The loop will exit because keyFile.available() becomes false...

TD-er replied:
Have you actually tried it?

The reason I'm asking is that I've seen some very odd results with `readStringUntil` in the past, though that was about 6 or 7 years ago...

chromoxdor replied:
I think i did but since i am unsure i'll do it again :)

chromoxdor replied:
I tested it without any issues

TD-er commented on src/src/Helpers/HTTPResponseParser.cpp:
```diff
@@ -0,0 +1,399 @@
+          } else if (element.is<const char *>()) {
+            csvOutput += String(element.as<const char *>());
+          } else {
+            csvOutput += "unknown";
```
Please use `F("unknown")`

TD-er commented on src/src/Helpers/HTTPResponseParser.cpp:
```diff
@@ -0,0 +1,399 @@
+          currentIndex++;
+
+          if (currentIndex < arraySize) {
+            csvOutput += ",";
```
Appending a single char should be done with `+= ',';`

TD-er commented on src/src/Helpers/HTTPResponseParser.cpp:
```diff
@@ -0,0 +1,399 @@
+          }
+        }
+      } else {
+        csvOutput += "unknown";
```
Ditto, please use flash string

TD-er commented on src/src/Helpers/HTTPResponseParser.cpp:
```diff
@@ -0,0 +1,399 @@
+        csvOutput += "unknown";
+      }
+    } else {
+      csvOutput += "null"; // Indicate missing value
```
Ditto, please use flash string

TD-er commented on src/src/Helpers/HTTPResponseParser.cpp:
```diff
@@ -0,0 +1,399 @@
+    } else {
+      csvOutput += "null"; // Indicate missing value
+    }
+    csvOutput += ",";
```
Ditto, append a char, not a C-type string.

chromoxdor committed:
Changes as requested

tonhuisman commented on docs/source/Rules/HTTPResponseParser.rst:
```diff
@@ -0,0 +1,239 @@
+The Thingspeak Parser
+---------------------
+
+* ``SendToHTTP`` generates an event with the response of a thingspeak request (https://de.mathworks.com/help/thingspeak/readlastfieldentry.html & https://de.mathworks.com/help/thingspeak/readdata.html)
```
These urls seem to redirect to a language specific version, depending on the user location, even though the content is English (AFAICS).
Maybe it's more universal if `www` is used instead of `de`.

chromoxdor replied:
true.. www seems to do the trick

tonhuisman commented on src/src/Helpers/HTTPResponseParser.cpp:
```diff
@@ -0,0 +1,397 @@
+               strformat(F("Response exceeds %d characters which could cause instabilities or crashes!"), RESPONSE_MAX_LENGTH));
+      }
+
+      auto processAndQueueParams = [](const String& url, const String& str, const String& eventName) {
+                                     // Extract the parameters from the URL
+                                     int start = url.indexOf(eventName + '=');
+
+                                     if (start == -1) {
+                                       return; // No parameters found for the given eventName
+                                     }
```
The formatting of this part of the source (and the next 75-odd lines) seems a bit weird, did Uncrustify format it this way?

chromoxdor replied:
I did format it with uncrustify... don`t know if it is supposed to be like that

tonhuisman replied:
May be triggered by the square brackets, don't think we've used that elsewhere.

tonhuisman commented on src/src/Helpers/HTTPResponseParser.cpp:
```diff
@@ -0,0 +1,397 @@
+      // Check if the URL ends with a number
+      if ((uri.length() > 0) && isDigit(uri.charAt(uri.length() - 1))) {
+        // Find the position of the last non-digit character
+        int i = uri.length() - 1;
```
Moving this `int i =...` declaration & assignment outside the `if` reduces the code somewhat, by not calculating `uri.length() - 1` twice, and allowing the condition to be somewhat simpler ...
`if (i > -1 && isDigit(uri.charAt(i))) {`

chromoxdor committed:
Update HTTPResponseParser.rst

tonhuisman commented on src/src/Helpers/HTTPResponseParser.cpp:
```diff
@@ -0,0 +1,397 @@
+    JsonVariant value = *root;
+    size_t start = 0, end;
+
+    while ((end = key.indexOf('.', start)) != (unsigned int)-1) {
+      String part = key.substring(start, end);
+      value = value[part];
```
`String` variables that won't change can be declared `const String`, so the compiler can often optimize them better.

tonhuisman commented on src/src/Helpers/HTTPResponseParser.cpp:
```diff
@@ -0,0 +1,397 @@
+
+    // Handle the last part of the key
+    if (!value.isNull()) {
+      String lastPart = key.substring(start);
```
Ditto

tonhuisman commented on src/src/Helpers/HTTPResponseParser.cpp:
```diff
@@ -0,0 +1,397 @@
+          if (element.is<int>()) {
+            csvOutput += String(element.as<int>());
+          } else if (element.is<float>()) {
+            csvOutput += doubleToString(element.as<double>(), nr_decimals, 1);
```
For boolean arguments it's 'nicer' (for readability) to pass true/false instead of 1/0

tonhuisman commented on src/src/Helpers/HTTPResponseParser.cpp:
```diff
@@ -0,0 +1,397 @@
+    }
+
+    // Log the results
+    addLog(LOG_LEVEL_INFO, strformat(F("Successfully processed %d out of %d keys"), successfullyProcessedCount, keyCount));
```
Adding a check `if (loglevelActiveFor(LOG_LEVEL_INFO))` avoids building the string if it won't get logged, especially on memory-tight units

chromoxdor committed:
Update HTTPResponseParser.cpp

TD-er commented:
Well if you can do a last-minute check to make sure it is still working after these changes, I will merge it when no issues found.
So please let me know if you did a quick test.

chromoxdor commented:
Check completed. 
Everything seems to work as expected...

chromoxdor commented:
Thanks :)

TD-er commented:
Uploaded the build including this to the web flasher: https://td-er.nl/ESPEasy/latest/
